Add benchmark_pub.py and benchmark_sub.cpp

mauropasse opened the pull request:
- Added new ROS 2 package `benchmark_msgs` with `BenchmarkData.msg` (std_msgs/Header + uint8[]).
- Implemented `benchmark_pub.py`: Python FlatPublisher to publish configurable size messages (sweep/duration modes).
- Implemented `benchmark_sub.cpp`: C++ FlatSubscriber that logs latency per message to `latency_log.csv`.
- Added `analyze_latency.py` script to plot and summarize latency results.
- Updated `flatros2/examples/CMakeLists.txt` to build/install new benchmark nodes.
- Updated `flatros2/package.xml` and `CMakeLists.txt` to depend on `benchmark_msgs`.
- Exposed `Flat` symbol in Python init for convenience.
- Significantly expanded README with:
  - Workspace layout/setup instructions
  - Troubleshooting tips
  - How to run the benchmark pub/sub nodes
  - How to analyze results with the new script
- Updated `iceoryx.repos` to include additional Jazzy dependencies: `common_interfaces`, `test_interface_files`.

mpasserino-rai committed:
Add FlatROS2 benchmarking tools, README guide, and analyze script

- Added new ROS 2 package `benchmark_msgs` with `BenchmarkData.msg` (std_msgs/Header + uint8[]).
- Implemented `benchmark_pub.py`: Python FlatPublisher to publish configurable size messages (sweep/duration modes).
- Implemented `benchmark_sub.cpp`: C++ FlatSubscriber that logs latency per message to `latency_log.csv`.
- Added `analyze_latency.py` script to plot and summarize latency results.
- Updated `flatros2/examples/CMakeLists.txt` to build/install new benchmark nodes.
- Updated `flatros2/package.xml` and `CMakeLists.txt` to depend on `benchmark_msgs`.
- Exposed `Flat` symbol in Python init for convenience.
- Significantly expanded README with:
  - Workspace layout/setup instructions
  - Troubleshooting tips
  - How to run the benchmark pub/sub nodes
  - How to analyze results with the new script
- Updated `iceoryx.repos` to include additional Jazzy dependencies: `common_interfaces`, `test_interface_files`.

mpasserino-rai committed:
Use spin() - Improve latency plot

mpasserino-rai committed:
Add more features to the pub/sub benchmark

hidmic commented on README.md:
```diff
@@ -43,17 +43,60 @@ The easiest way to get started is by using the provided Visual Studio Code Dev C
+- For the ROS 2 core workspace, use the provided `ros2-core.repos` file:
+
+This step is optional! Is just to have the source code be able
+to instrument code and debug better.
```
@mauropasse nit: 
```suggestion
This step is optional! It is just to have the source code available for instrumentation and debug.
```

hidmic commented on README.md:
````diff
@@ -43,17 +43,60 @@ The easiest way to get started is by using the provided Visual Studio Code Dev C
+
+---
+
+## Troubleshooting & Setup Tips
+
+If you encounter build or permission errors, try the following:
+
+**Fix permissions so colcon can write logs/builds:**
+
+```bash
+sudo chown -R developer:ekumen .
+```
+
+**Fix missing ACL headers needed by iceoryx:**
+
+```bash
+sudo apt update && sudo apt install -y libacl1-dev
+```
````
@mauropasse meta: this means the container needs some work but OK with docs for now.

hidmic commented on README.md:
```diff
@@ -103,8 +146,86 @@ ros2 run flatros2 image_viewer_node.py --ros-args -r image:=image/edges
 
 This project is licensed under the Apache 2.0 License - see the [LICENSE](./LICENSE) file for details.
+
+## Benchmarking Flatros2 Performance
```
@mauropasse nit^2: consider moving this to a BENCHMARKING.md file.

hidmic commented on benchmark_msgs/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+find_package(std_msgs REQUIRED)
+
+rosidl_generate_interfaces(${PROJECT_NAME}
+  "msg/BenchmarkData.msg"
```
@mauropasse out of curiosity, why not [`std_msgs/msg/UInt8MultiArray`](https://docs.ros.org/en/noetic/api/std_msgs/html/msg/UInt8MultiArray.html) or similar?

mauropasse replied:
Good question! I considered `std_msgs/msg/UInt8MultiArray`, but it doesn’t include a `Header`, so not possible to attach timestamps. I needed something I could timestamp for the latency calculation.

hidmic commented on flatros2/examples/CMakeLists.txt:
```diff
@@ -1,16 +1,39 @@
 
-add_executable(face_detector_node face_detector_node.cpp)
-target_link_libraries(face_detector_node PRIVATE 
+add_executable(edge_detector_node edge_detector_node.cpp)
```
@mauropasse :heart:

hidmic commented on flatros2/examples/benchmark_pub.py:
```diff
@@ -0,0 +1,206 @@
+        self.publisher.publish_loaned_message(msg)
+        t1 = time.perf_counter()
+        # print(f"{self.GREEN}Published {self.sizes[self.index]} bytes in {self.YELLOW}{1e3*(t1 - t0):.2f}{self.RESET} ms")
+        time.sleep(self.period_ms / 1000.0)
```
@mauropasse isn't this a timer callback :thinking: ? Why sleep? Same below.

hidmic reviewed (approved):
Looks good. What numbers are we getting?

mpasserino-rai committed:
fix: use timer period from --period argument and remove unnecessary sleeps

mpasserino-rai committed:
docs: move benchmarking instructions to BENCHMARKING.md

- Extracted the benchmarking section from README.md into a new BENCHMARKING.md file for clarity.
- README.md now references only core usage and setup instructions.
- No content changes to the benchmarking instructions themselves.

mpasserino-rai committed:
Expand doc showing how to get CPU Flamegraphs

mpasserino-rai committed:
Pub: Only take msg sizes specified thorugh args

mpasserino-rai committed:
Add fixed-size benchmark messages

mpasserino-rai committed:
feat: add fixed-size benchmark pub/sub nodes using new Stamped* messages

- Add C++ publisher node (benchmark_pub_fixed_size.cpp) that publishes fixed-size messages
  using Stamped10B, Stamped100KB, Stamped1MB, Stamped2MB, and Stamped4MB.

- Add C++ subscriber node (benchmark_sub_fixed_size.cpp) that listens on the corresponding topics,
  logs latency to a CSV file, and optionally prints debug info.

- Register both executables in examples/CMakeLists.txt for build and install.

- These nodes mirror the functionality of the Python benchmark_pub.py but use statically sized
  message types to allow comparison with flatros2 zero-copy variable-size benchmarks.

mpasserino-rai committed:
Add ros2-core.repos

hidmic commented on benchmark_msgs/msg/Stamped1MB.msg:
```diff
@@ -0,0 +1,3 @@
+# This message contains a header and a fixed-size array of 1,000,000 bytes
+std_msgs/Header header
+uint8[1000000] data
```
@mauropasse I don't follow, what's the point of fixed size messages?

hidmic commented on iceoryx.repos:
```diff
@@ -9,5 +10,15 @@ repositories:
+    url: https://github.com/Ekumen-OS/rmw_iceoryx2.git
     version: flatros2-compatible
+
+  # Needed to build flatros
```
@mauropasse why not add it to the `package.xml` then?

hidmic commented:
This has been superseded by #2.
